### DISCUSSION TITLE
How to trigger Select.Changed when reset the value of select field?

### DISCUSSION THREAD

**Discussion by Tim-Cao** [Textualize/textual#3400] [Q&A] [answered]
```
with VerticalScroll(id="nav-5"):
    DESCRIPTION_TYPE = [
        ("Steps to reproduce", "STR"),
        ("Test Cases", "TC"),
    ]

    yield Label("Enter the Jira ticket number: ")
    yield Input(id="jira-ticket-number-5")
    yield Label("Select the description type: ")
    yield Select(DESCRIPTION_TYPE, id="description-type")
    yield Label(
        "Enter the steps to reproduce: ",
        classes="unselected",
        id="steps-to-reproduce-label",
    )
    yield TextArea(STR, classes="unselected", id="steps-to-reproduce")
    yield Static()
    yield Button("Submit", variant="primary", id="button-5")


def write_description(self) -> None:
    ticket_number = self.query_one("#jira-ticket-number-5").value
    type = self.query_one("#description-type").value
    steps_to_reproduce = self.query_one("#steps-to-reproduce").text

    self.query_one("#button-5").disabled = True
    self.query_one(RichLog).clear()
    self.query_one(RichLog).begin_capture_print()

    write_description.main(ticket_number, type, steps_to_reproduce)

    self.query_one("#button-5").disabled = False
    self.query_one("#jira-ticket-number-5").value = ""
    self.query_one("#description-type").value = ""


def on_select_changed(self, event: Select.Changed) -> None:
    self.query_one("#steps-to-reproduce").set_class(event.value == "STR", "visible")
    self.query_one("#steps-to-reproduce-label").set_class(
        event.value == "STR", "visible"
    )
```

I would like to reset the classes of #steps-to-reproduce and #steps-to-reproduce-label when select #description-type is reset.

**Comment by davep** [comment]
It's not every clear from your question what you're actually asking for help with; the subject of the question doesn't seem to quite match the thing that's asked in the body and the code that's posted isn't really that helpful. Could I encourage you to reword your question, making it as clear as possible what you're looking for help with, and if code will help [try and make it a MRE](https://github.com/Textualize/textual/discussions/3361).

**Reply by Tim-Cao** [reply]
Thanks for your suggestion. @davep 
```
from textual import on, work
from textual.app import App, ComposeResult, events
from textual.containers import *
from textual.widgets import *


class ScriptApp(App):
    CSS = """
        TextArea {
            height: 20;
        }

        .unselected {
            display: none;
        }

        .unselected.visible {
            display: block;
        }
    """

    def compose(self) -> ComposeResult:
        STR = "(on) *Steps to reproduce:*\n"

        DESCRIPTION_TYPE = [
            ("Steps to reproduce", "STR"),
            ("Test Cases", "TC"),
        ]

        yield Label("Select the description type: ")
        yield Select(DESCRIPTION_TYPE, id="description-type")
        yield Label(
            "Enter the steps to reproduce: ",
            classes="unselected",
            id="steps-to-reproduce-label",
        )
        yield TextArea(STR, classes="unselected", id="steps-to-reproduce")
        yield Static()
        yield Button("Submit", variant="primary", id="button-5")

    @work(exclusive=True, thread=True)
    def write_description(self) -> None:
        self.query_one("#description-type").value
        self.query_one("#steps-to-reproduce").text

        self.query_one("#button-5").disabled = False
        self.query_one("#description-type").value = ""

    def on_button_pressed(self, event: Button.Pressed) -> None:
        if event.button.id == "button-5":
            self.write_description()

    def on_select_changed(self, event: Select.Changed) -> None:
        self.query_one("#steps-to-reproduce").set_class(event.value == "STR", "visible")
        self.query_one("#steps-to-reproduce-label").set_class(
            event.value == "STR", "visible"
        )


app = ScriptApp()

if __name__ == "__main__":
    app.run()
```
Here is an ootb example.

**Reply by Tim-Cao** [reply]
I would like to reset the classes of #steps-to-reproduce and #steps-to-reproduce-label when select #description-type is reset. When the Select the description type is cleared, how to trigger Select.Changed to hide #steps-to-reproduce and #steps-to-reproduce-label?

**Reply by Tim-Cao** [reply]
https://github.com/Textualize/textual/assets/52661397/dd3a3fda-ad50-42ea-9181-0734497c67cc

**Reply by TomJGooding** [reply]
> when reset the value of select field

```python
 self.query_one(Select).value = None
```

When the value of a Select widget is "reset" above, has the select value ***changed***?

I don't understand why you would need anything to trigger this, but your code is difficult to follow.

**Reply by Tim-Cao** [reply]
Hi @TomJGooding,
The value of select is reset. But it didn't trigger Select.Changed. If it trigger Select.Changed, the TextArea should be hidden again.
Imagine this is a form, when submit successfully, I would like the form reset to original prepare for the next time.

**Reply by davep** [reply]
So, is the question here: "should assignment to `Select.value` cause a `Select.Changed` event to be posted"? If so, then no, it's my understanding that `Select`, and a number of other widgets that have some sort of message that indicates a state-change, only post those messages when the *user* makes the change. Changes made by code don't and it's expected that the developer will know they're making the change and act accordingly.

So, for example, I think I'd have a method for clearing the values, and then when I want to clear those values in code I'd call that method; and in the handler for `Select.Changed` I'd see if the value is `None` and also call that clearing method.

**Reply by TomJGooding** [reply]
@davep Hm, that wasn't my understanding of messages! For example see [this test](https://github.com/Textualize/textual/blob/main/tests/test_message_pump.py#L92) where assigning the value will increment the changed hits.

**Reply by davep** [reply]
@TomJGooding You'd expect this to add things to the log with no input from you, in that case, right?

```python
from itertools import cycle

from textual import on
from textual.app import App, ComposeResult
from textual.widgets import Log, Select

class SelectMessagesApp(App[None]):

    def compose(self) -> ComposeResult:
        yield Select[int](
            (("One", 0), ("Two", 1))
        )
        yield Log()

    def on_mount(self) -> None:
        self.selections = cycle((0, 1, None))
        self.set_interval(0.25, self.change_value)

    @on(Select.Changed)
    def log_select(self, event: Select.Changed) -> None:
        self.query_one(Log).write_line(f"{event!r} -> {event.value!r}")

    def change_value(self) -> None:
        self.query_one(Select).value = next(self.selections)

if __name__ == "__main__":
    SelectMessagesApp().run()
```

**Reply by Tim-Cao** [reply, marked as answer]
The correct way is to remove_class? @davep
```
        self.query_one("#steps-to-reproduce").remove_class("visible")
```

**Reply by davep** [reply]
You can confirm this [in the Textual docs](https://textual.textualize.io/guide/CSS/#class-name-selector).

**Reply by TomJGooding** [reply]
@davep I would expect `Select.Changed` to behave like this message in other widgets, so actually @Tim-Cao I owe you an apology.

Here's just a few examples below. I understand if the Select message is different by design, but if that's the case perhaps this deserves a note in the docs.

<details>
<summary>Input.Changed</summary>

```python
from itertools import cycle

from textual import on
from textual.app import App, ComposeResult
from textual.widgets import Input, Log


class MessagesApp(App[None]):
    def compose(self) -> ComposeResult:
        yield Input()
        yield Log()

    def on_mount(self) -> None:
        self.text = cycle(("foo", "bar", "baz"))
        self.set_interval(0.25, self.change_value)

    @on(Input.Changed)
    def log_select(self, event: Input.Changed) -> None:
        self.query_one(Log).write_line(f"{event!r} -> {event.value!r}")

    def change_value(self) -> None:
        self.query_one(Input).value = next(self.text)


if __name__ == "__main__":
    MessagesApp().run()
```

</details>

<details>
<summary>Switch.Changed</summary>

```python
from textual import on
from textual.app import App, ComposeResult
from textual.widgets import Log, Switch


class MessagesApp(App[None]):
    def compose(self) -> ComposeResult:
        yield Switch()
        yield Log()

    def on_mount(self) -> None:
        self.set_interval(0.25, self.change_value)

    @on(Switch.Changed)
    def log_select(self, event: Switch.Changed) -> None:
        self.query_one(Log).write_line(f"{event!r} -> {event.value!r}")

    def change_value(self) -> None:
        switch = self.query_one(Switch)
        switch.value = not switch.value


if __name__ == "__main__":
    MessagesApp().run()
```

</details>

<details>
<summary>Checkbox.Changed</summary>

```python
from textual import on
from textual.app import App, ComposeResult
from textual.widgets import Checkbox, Log


class MessagesApp(App[None]):
    def compose(self) -> ComposeResult:
        yield Checkbox("Foo")
        yield Log()

    def on_mount(self) -> None:
        self.set_interval(0.25, self.change_value)

    @on(Checkbox.Changed)
    def log_select(self, event: Checkbox.Changed) -> None:
        self.query_one(Log).write_line(f"{event!r} -> {event.value!r}")

    def change_value(self) -> None:
        checkbox = self.query_one(Checkbox)
        checkbox.value = not checkbox.value


if __name__ == "__main__":
    MessagesApp().run()
```
</details>

**Reply by davep** [reply]
@TomJGooding Based on previous conversations I've been involved in, I think the design choice for Textual is that it's those widgets that are behaving "incorrectly" (for what it's worth my own personal taste is they're doing it correctly; but I don't get final say on that 😉).

I think you raise a good point and it's issue-worthy that one approach or another is settled on and applied across all widgets with such behaviour.

**Reply by TomJGooding** [reply]
>  (for what it's worth my own personal taste is they're doing it correctly; but I don't get final say on that 😉).

Also for what it's worth, I agree!

> I think you raise a good point and it's issue-worthy that one approach or another is settled on and applied across all widgets with such behaviour.

Should I create a new issue? I'm just checking we aren't doubling up and perhaps it would better coming from you?

**Reply by davep** [reply]
Go for it. :-)

**Reply by Tim-Cao** [reply]
Hi @davep and @TomJGooding, do you know when will this request for `Select.Changed` be released?

**Reply by davep** [reply]
You can always see what's being worked on [by looking at the PRs](https://github.com/Textualize/textual/pulls); we don't have a specific release schedule for anything though, features or specific versions of Textual.
